[Android/tflite] Failed to invoke tflite model with nnapi.

The test `testInvokeDynamicTFLite()` keeps failing by crash.
(https://github.com/nnstreamer/nnstreamer/blob/for-next/api/android/api/src/androidTest/java/org/nnsuite/nnstreamer/APITestSingleShot.java#L272)

Currently, tflite (1.13.1) utilizes nnapi by default: `use_nnapi = TRUE;`
(https://github.com/nnstreamer/nnstreamer/blob/master/ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc#L574)

If I modify this line with `use_nnapi = FALSE;`, the failures do not occur.
This may be related with file access permission.

Test failure log:
```
...
2020-09-21 16:03:25.046 4220-4323/? D/nnstreamer: Framework = tensorflow-lite
2020-09-21 16:03:25.046 4220-4323/? I/GLib: nnapi = 1, accl = cpu
2020-09-21 16:03:25.047 684-684/? D/keymaster_swd: keymaster_swd [WRN] (swd_run_cb:249) swd_finish() returns 0
2020-09-21 16:03:25.047 4220-4323/? W/nnstreamer: The number of tensor name is mismatched in filter.
2020-09-21 16:03:25.047 4220-4323/? W/nnstreamer: The number of tensor name is mismatched in filter.
2020-09-21 16:03:25.047 1053-4566/? D/keystore: [FINISH::end]
2020-09-21 16:03:25.048 4220-4323/? E/NeuralNetworks: SecFeature = 3
2020-09-21 16:03:25.050 4220-4323/? D/NativeHcmServiceManager: [NativeHcm] BpSemHcmManager::BpSemHcmManager()
2020-09-21 16:03:25.050 1055-12636/? I/HcmInfo::HcmManagerService: ctc pid=4220, uid=10131, Package Name:org.nnsuite.nnstreamer.test
2020-09-21 16:03:25.051 1055-2553/? D/CustomFrequencyManagerService: acquireDVFSLockLocked : type : DVFS_MIN_LIMIT  frequency : 2457600  uid : 1000  pid : 1055  pkgName : hcm@81
2020-09-21 16:03:25.060 4220-4323/? I/TypeManager: Failed to read /vendor/etc/nnapi_extensions_app_allowlist ; No app allowlisted for vendor extensions use.
2020-09-21 16:03:25.061 4220-4323/? E/nnstreamer.tes: getSupportedOperations_1_2 failure: Status(EX_TRANSACTION_FAILED): 'FAILED_TRANSACTION: '
2020-09-21 16:03:25.061 4220-4323/? E/Manager: IDevice::getSupportedOperations returned the error GENERAL_FAILURE
2020-09-21 16:03:25.061 4220-4323/? E/nnstreamer.tes: getSupportedOperations_1_2 failure: Status(EX_TRANSACTION_FAILED): 'FAILED_TRANSACTION: '
2020-09-21 16:03:25.061 4220-4323/? E/Manager: IDevice::getSupportedOperations returned the error GENERAL_FAILURE
2020-09-21 16:03:25.061 747-747/? E/audit: type=1400 audit(1600671805.056:101959): avc:  denied  { read } for  pid=4220 comm="roidJUnitRunner" path="/storage/emulated/0/nnstreamer/test/add/add.tflite" dev="sdcardfs" ino=23663 scontext=u:r:hal_neuralnetworks_default:s0 tcontext=u:object_r:sdcardfs:s0 tclass=file permissive=0 SEPF_SM-G981N_10_0021 audit_filtered
2020-09-21 16:03:25.061 4220-4323/? E/nnstreamer.tes: getSupportedOperations_1_2 failure: Status(EX_TRANSACTION_FAILED): 'FAILED_TRANSACTION: '
2020-09-21 16:03:25.061 747-747/? E/audit: type=1300 audit(1600671805.056:101959): arch=c00000b7 syscall=29 success=yes exit=0 a0=43 a1=c0306201 a2=7c679f03a0 a3=7c679f06a0 items=0 ppid=765 pid=4220 auid=4294967295 uid=10131 gid=10131 euid=10131 suid=10131 fsuid=10131 egid=10131 sgid=10131 fsgid=10131 tty=(none) ses=4294967295 comm="roidJUnitRunner" exe="/system/bin/app_process64" subj=u:r:untrusted_app_27:s0:c131,c256,c512,c768 key=(null)
2020-09-21 16:03:25.061 4220-4323/? E/Manager: IDevice::getSupportedOperations returned the error GENERAL_FAILURE
2020-09-21 16:03:25.061 747-747/? E/audit: type=1327 audit(1600671805.056:101959): proctitle="org.nnsuite.nnstreamer.test"
2020-09-21 16:03:25.061 4220-4323/? E/nnstreamer.tes: getSupportedOperations_1_2 failure: Status(EX_TRANSACTION_FAILED): 'FAILED_TRANSACTION: '
2020-09-21 16:03:25.062 4220-4323/? E/Manager: IDevice::getSupportedOperations returned the error GENERAL_FAILURE
2020-09-21 16:03:25.062 747-747/? E/audit: type=1400 audit(1600671805.056:101960): avc:  denied  { read } for  pid=4220 comm="roidJUnitRunner" path="/storage/emulated/0/nnstreamer/test/add/add.tflite" dev="sdcardfs" ino=23663 scontext=u:r:hal_neuralnetworks_default:s0 tcontext=u:object_r:sdcardfs:s0 tclass=file permissive=0 SEPF_SM-G981N_10_0021 audit_filtered
2020-09-21 16:03:25.062 747-747/? E/audit: type=1300 audit(1600671805.056:101960): arch=c00000b7 syscall=29 success=yes exit=0 a0=43 a1=c0306201 a2=7c679f03a0 a3=7c679f06a0 items=0 ppid=765 pid=4220 auid=4294967295 uid=10131 gid=10131 euid=10131 suid=10131 fsuid=10131 egid=10131 sgid=10131 fsgid=10131 tty=(none) ses=4294967295 comm="roidJUnitRunner" exe="/system/bin/app_process64" subj=u:r:untrusted_app_27:s0:c131,c256,c512,c768 key=(null)
2020-09-21 16:03:25.062 747-747/? E/audit: type=1327 audit(1600671805.056:101960): proctitle="org.nnsuite.nnstreamer.test"
2020-09-21 16:03:25.062 747-747/? E/audit: type=1400 audit(1600671805.056:101961): avc:  denied  { read } for  pid=4220 comm="roidJUnitRunner" path="/storage/emulated/0/nnstreamer/test/add/add.tflite" dev="sdcardfs" ino=23663 scontext=u:r:hal_neuralnetworks_default:s0 tcontext=u:object_r:sdcardfs:s0 tclass=file permissive=0 SEPF_SM-G981N_10_0021 audit_filtered

...

2020-09-21 16:03:25.093 4220-4323/? E/ExecutionBuilder: Setting argument with invalid length: 8, expected length: 4
2020-09-21 16:03:25.093 4220-4323/? E/tflite: Aborting since NNAPI returned failure nnapi_delegate.cc:805

...

2020-09-21 16:03:25.111 1055-11453/? I/ActivityManager: Process org.nnsuite.nnstreamer.test (pid 4220) has died: fore FGS (2707,4407)
2020-09-21 16:03:25.112 1055-11453/? W/ActivityManager: Crash of app org.nnsuite.nnstreamer.test running instrumentation ComponentInfo{org.nnsuite.nnstreamer.test/android.support.test.runner.AndroidJUnitRunner}
2020-09-21 16:03:25.112 1055-11453/? I/ActivityManager: Force stopping org.nnsuite.nnstreamer.test appid=10131 user=0: finished inst
```

With tflite v1.15.2, the test does not crash but give unexpected return value:
expected: 3.5f, but was : 0.0f

```
2020-09-24 14:07:09.586 6035-6141/? I/nnstreamer: NNStreamer 1.7.0 GStreamer 1.16.2 GLib 2.56.1
2020-09-24 14:07:09.588 6035-6141/? D/nnstreamer: Framework = tensorflow-lite
2020-09-24 14:07:09.588 6035-6141/? I/GLib: nnapi = 1, accl = cpu
2020-09-24 14:07:09.589 6035-6141/? I/tflite: Initialized TensorFlow Lite runtime.
2020-09-24 14:07:09.589 6035-6141/? W/nnstreamer: The number of tensor name is mismatched in filter.
2020-09-24 14:07:09.589 6035-6141/? W/nnstreamer: The number of tensor name is mismatched in filter.
2020-09-24 14:07:09.589 6035-6141/? I/tflite: Created TensorFlow Lite delegate for NNAPI.
...
2020-09-24 14:07:09.593 6035-6141/? I/Manager: Capab {.relaxedFloat32toFloat16PerformanceScalar = {.execTime = 0.600000, .powerUsage = 0.800000}, .relaxedFloat32toFloat16PerformanceTensor = {.execTime = 0.600000, .powerUsage = 0.800000}, .operandPerformance = [16]{{.type = FLOAT32, .info = {.execTime = 0.600000, .powerUsage = 0.800000}}, {.type = INT32, .info = {.execTime = 0.600000, .powerUsage = 0.800000}}, {.type = UINT32, .info = {.execTime = 0.600000, .powerUsage = 0.800000}}, {.type = TENSOR_FLOAT32, .info = {.execTime = 0.600000, .powerUsage = 0.800000}}, {.type = TENSOR_INT32, .info = {.execTime = 0.600000, .powerUsage = 0.800000}}, {.type = TENSOR_QUANT8_ASYMM, .info = {.execTime = 0.700000, .powerUsage = 0.700000}}, {.type = BOOL, .info = {.execTime = 0.600000, .powerUsage = 0.800000}}, {.type = TENSOR_QUANT16_SYMM, .info = {.execTime = 0.700000, .powerUsage = 0.700000}}, {.type = TENSOR_FLOAT16, .info = {.execTime = 0.600000, .powerUsage = 0.800000}}, {.type = TENSOR_BOOL8, .info = {.execTime = 0.600000, .powerUsage = 0.800000}}, {.type = FLOA
2020-09-24 14:07:09.593 6035-6141/? I/Manager: Found interface qti-dsp
2020-09-24 14:07:09.595 6035-6141/? I/Manager: Capab {.relaxedFloat32toFloat16PerformanceScalar = {.execTime = 2.000000, .powerUsage = 2.000000}, .relaxedFloat32toFloat16PerformanceTensor = {.execTime = 2.000000, .powerUsage = 2.000000}, .operandPerformance = [16]{{.type = FLOAT32, .info = {.execTime = 2.000000, .powerUsage = 2.000000}}, {.type = INT32, .info = {.execTime = 2.000000, .powerUsage = 2.000000}}, {.type = UINT32, .info = {.execTime = 2.000000, .powerUsage = 2.000000}}, {.type = TENSOR_FLOAT32, .info = {.execTime = 2.000000, .powerUsage = 2.000000}}, {.type = TENSOR_INT32, .info = {.execTime = 2.000000, .powerUsage = 2.000000}}, {.type = TENSOR_QUANT8_ASYMM, .info = {.execTime = 1.100000, .powerUsage = 1.100000}}, {.type = BOOL, .info = {.execTime = 2.000000, .powerUsage = 2.000000}}, {.type = TENSOR_QUANT16_SYMM, .info = {.execTime = 1.100000, .powerUsage = 1.100000}}, {.type = TENSOR_FLOAT16, .info = {.execTime = 2.000000, .powerUsage = 2.000000}}, {.type = TENSOR_BOOL8, .info = {.execTime = 2.000000, .powerUsage = 2.000000}}, {.type = FLOA
2020-09-24 14:07:09.595 6035-6141/? I/Manager: Found interface qti-gpu
2020-09-24 14:07:09.596 6035-6141/? I/Manager: Capab {.relaxedFloat32toFloat16PerformanceScalar = {.execTime = 1.100000, .powerUsage = 1.100000}, .relaxedFloat32toFloat16PerformanceTensor = {.execTime = 1.100000, .powerUsage = 1.100000}, .operandPerformance = [16]{{.type = FLOAT32, .info = {.execTime = 1.100000, .powerUsage = 1.100000}}, {.type = INT32, .info = {.execTime = 1.100000, .powerUsage = 1.100000}}, {.type = UINT32, .info = {.execTime = 1.100000, .powerUsage = 1.100000}}, {.type = TENSOR_FLOAT32, .info = {.execTime = 1.100000, .powerUsage = 1.100000}}, {.type = TENSOR_INT32, .info = {.execTime = 1.100000, .powerUsage = 1.100000}}, {.type = TENSOR_QUANT8_ASYMM, .info = {.execTime = 2.000000, .powerUsage = 2.000000}}, {.type = BOOL, .info = {.execTime = 1.100000, .powerUsage = 1.100000}}, {.type = TENSOR_QUANT16_SYMM, .info = {.execTime = 2.000000, .powerUsage = 2.000000}}, {.type = TENSOR_FLOAT16, .info = {.execTime = 1.100000, .powerUsage = 1.100000}}, {.type = TENSOR_BOOL8, .info = {.execTime = 1.100000, .powerUsage = 1.100000}}, {.type = FLOA
2020-09-24 14:07:09.596 6035-6141/? I/Manager: Found interface qti-hta
2020-09-24 14:07:09.599 6035-6141/? I/Manager: Capab {.relaxedFloat32toFloat16PerformanceScalar = {.execTime = 2.000000, .powerUsage = 2.000000}, .relaxedFloat32toFloat16PerformanceTensor = {.execTime = 2.000000, .powerUsage = 2.000000}, .operandPerformance = [16]{{.type = FLOAT32, .info = {.execTime = 2.000000, .powerUsage = 2.000000}}, {.type = INT32, .info = {.execTime = 2.000000, .powerUsage = 2.000000}}, {.type = UINT32, .info = {.execTime = 2.000000, .powerUsage = 2.000000}}, {.type = TENSOR_FLOAT32, .info = {.execTime = 2.000000, .powerUsage = 2.000000}}, {.type = TENSOR_INT32, .info = {.execTime = 2.000000, .powerUsage = 2.000000}}, {.type = TENSOR_QUANT8_ASYMM, .info = {.execTime = 1.100000, .powerUsage = 1.100000}}, {.type = BOOL, .info = {.execTime = 2.000000, .powerUsage = 2.000000}}, {.type = TENSOR_QUANT16_SYMM, .info = {.execTime = 1.100000, .powerUsage = 1.100000}}, {.type = TENSOR_FLOAT16, .info = {.execTime = 2.000000, .powerUsage = 2.000000}}, {.type = TENSOR_BOOL8, .info = {.execTime = 2.000000, .powerUsage = 2.000000}}, {.type = FLOA
2020-09-24 14:07:09.600 6035-6141/? E/NeuralNetworks: SecFeature = 3
2020-09-24 14:07:09.601 6035-6141/? D/NativeHcmServiceManager: [NativeHcm] BpSemHcmManager::BpSemHcmManager()
2020-09-24 14:07:09.602 1055-10282/? I/HcmInfo::HcmManagerService: ctc pid=6035, uid=10460, Package Name:org.nnsuite.nnstreamer.test
2020-09-24 14:07:09.602 6035-6141/? I/TypeManager: Failed to read /vendor/etc/nnapi_extensions_app_allowlist ; No app allowlisted for vendor extensions use.
2020-09-24 14:07:09.575 1055-12252/? I/GenerationRegistry: mBackingStore.isClosed() : false
2020-09-24 14:07:09.602 1055-2553/? D/CustomFrequencyManagerService: acquireDVFSLockLocked : type : DVFS_MIN_LIMIT  frequency : 2457600  uid : 1000  pid : 1055  pkgName : hcm@81
2020-09-24 14:07:09.604 6035-6141/? E/nnstreamer.tes: getSupportedOperations_1_2 failure: Status(EX_TRANSACTION_FAILED): 'FAILED_TRANSACTION: '
2020-09-24 14:07:09.604 6035-6141/? E/Manager: IDevice::getSupportedOperations returned the error GENERAL_FAILURE
2020-09-24 14:07:09.604 6035-6141/? E/nnstreamer.tes: getSupportedOperations_1_2 failure: Status(EX_TRANSACTION_FAILED): 'FAILED_TRANSACTION: '
2020-09-24 14:07:09.604 6035-6141/? E/Manager: IDevice::getSupportedOperations returned the error GENERAL_FAILURE
2020-09-24 14:07:09.604 747-747/? E/audit: type=1400 audit(1600924029.602:144890): avc:  denied  { read } for  pid=6035 comm="roidJUnitRunner" path="/storage/emulated/0/nnstreamer/test/add/add.tflite" dev="sdcardfs" ino=23663 scontext=u:r:hal_neuralnetworks_default:s0 tcontext=u:object_r:sdcardfs:s0 tclass=file permissive=0 SEPF_SM-G981N_10_0021 audit_filtered
2020-09-24 14:07:09.604 6035-6141/? E/nnstreamer.tes: getSupportedOperations_1_2 failure: Status(EX_TRANSACTION_FAILED): 'FAILED_TRANSACTION: '
2020-09-24 14:07:09.604 6035-6141/? E/Manager: IDevice::getSupportedOperations returned the error GENERAL_FAILURE
2020-09-24 14:07:09.604 747-747/? E/audit: type=1300 audit(1600924029.602:144890): arch=c00000b7 syscall=29 success=yes exit=0 a0=42 a1=c0306201 a2=7c679eff70 a3=7c679f0270 items=0 ppid=765 pid=6035 auid=4294967295 uid=10460 gid=10460 euid=10460 suid=10460 fsuid=10460 egid=10460 sgid=10460 fsgid=10460 tty=(none) ses=4294967295 comm="roidJUnitRunner" exe="/system/bin/app_process64" subj=u:r:untrusted_app_27:s0:c204,c257,c512,c768 key=(null)
2020-09-24 14:07:09.604 747-747/? E/audit: type=1327 audit(1600924029.602:144890): proctitle="org.nnsuite.nnstreamer.test"
2020-09-24 14:07:09.604 6035-6141/? E/nnstreamer.tes: getSupportedOperations_1_2 failure: Status(EX_TRANSACTION_FAILED): 'FAILED_TRANSACTION: '
2020-09-24 14:07:09.604 6035-6141/? E/Manager: IDevice::getSupportedOperations returned the error GENERAL_FAILURE
2020-09-24 14:07:09.604 747-747/? E/audit: type=1400 audit(1600924029.602:144891): avc:  denied  { read } for  pid=6035 comm="roidJUnitRunner" path="/storage/emulated/0/nnstreamer/test/add/add.tflite" dev="sdcardfs" ino=23663 scontext=u:r:hal_neuralnetworks_default:s0 tcontext=u:object_r:sdcardfs:s0 tclass=file permissive=0 SEPF_SM-G981N_10_0021 audit_filtered
2020-09-24 14:07:09.604 747-747/? E/audit: type=1300 audit(1600924029.602:144891): arch=c00000b7 syscall=29 success=yes exit=0 a0=42 a1=c0306201 a2=7c679eff70 a3=7c679f0270 items=0 ppid=765 pid=6035 auid=4294967295 uid=10460 gid=10460 euid=10460 suid=10460 fsuid=10460 egid=10460 sgid=10460 fsgid=10460 tty=(none) ses=4294967295 comm="roidJUnitRunner" exe="/system/bin/app_process64" subj=u:r:untrusted_app_27:s0:c204,c257,c512,c768 key=(null)
2020-09-24 14:07:09.605 747-747/? E/audit: type=1327 audit(1600924029.602:144891): proctitle="org.nnsuite.nnstreamer.test"
2020-09-24 14:07:09.605 6035-6141/? I/TestRunner: failed: testInvokeDynamicTFLite(org.nnsuite.nnstreamer.APITestSingleShot)
2020-09-24 14:07:09.605 6035-6141/? I/TestRunner: ----- begin exception -----
2020-09-24 14:07:09.605 6035-6141/? I/TestRunner: java.lang.AssertionError: expected:<3.5> but was:<0.0>
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.failNotEquals(Assert.java:834)
        at org.junit.Assert.assertEquals(Assert.java:575)
        at org.junit.Assert.assertEquals(Assert.java:700)
        at org.nnsuite.nnstreamer.APITestSingleShot.runInvokeDynamic(APITestSingleShot.java:259)
        at org.nnsuite.nnstreamer.APITestSingleShot.testInvokeDynamicTFLite(APITestSingleShot.java:278)
...
2020-09-24 14:07:09.605 6035-6141/? I/TestRunner: ----- end exception -----
2020-09-24 14:07:09.606 6035-6141/? I/TestRunner: finished: testInvokeDynamicTFLite(org.nnsuite.nnstreamer.APITestSingleShot)

```

--- a/ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc
+++ b/ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc
@@ -121,6 +121,9 @@ public:
   /** @brief cache input and output tensor ptr before invoke */
   int cacheInOutTensorPtr ();
 
+  /** @brief set delegate for the tflite interpreter */
+  void setDelegate (TfLiteDelegate* delegate) { delegate_ = delegate; }
+
 private:
   GMutex mutex;
   char *model_path;
@@ -138,6 +141,12 @@ private:
   int getTensorDim (int tensor_idx, tensor_dim dim);
   int setTensorProp (const std::vector<int> &tensor_idx_list,
       GstTensorsInfo * tensorMeta);
+
+  TfLiteDelegate* delegate_ = nullptr; /**< The delegate for tflite interpreter */
+
+#if defined(__ANDROID__) && (TFLITE_VERSION_MAJOR >= 2 || TFLITE_VERSION_MINOR >= 14)
+  std::unique_ptr<tflite::StatefulNnApiDelegate> stateful_nnapi_delegate; /**< The pointer of NNAPI delegate */
+#endif
 };
 
 /**
@@ -288,9 +297,6 @@ TFLiteInterpreter::loadModel (int num_threads, accl_hw accelerator)
     return -2;
   }
 
-  if (accelerator == ACCL_AUTO || accelerator == ACCL_GPU ||
-          accelerator == ACCL_NPU)
-    interpreter->UseNNAPI (TRUE);
   if (num_threads > 0) {
     int n = static_cast<int> (std::thread::hardware_concurrency ());
 
@@ -299,10 +305,29 @@ TFLiteInterpreter::loadModel (int num_threads, accl_hw accelerator)
     interpreter->SetNumThreads (n);
   }
 
-  if (interpreter->AllocateTensors () != kTfLiteOk) {
-    ml_loge ("Failed to allocate tensors\n");
-    return -2;
+  /** set nnapi delegate when accelerator set to auto (cpu.neon in Android), GPU or NPU */
+  if (accelerator == ACCL_CPU_NEON || accelerator == ACCL_GPU ||
+          accelerator == ACCL_NPU) {
+#if defined(__ANDROID__) && (TFLITE_VERSION_MAJOR >= 2 || TFLITE_VERSION_MINOR >= 14)
+    stateful_nnapi_delegate.reset (new tflite::StatefulNnApiDelegate ());
+    setDelegate (stateful_nnapi_delegate.get ());
+#else
+    ml_logw ("NNAPI delegate is available only in Android with tflite v1.14.0 or higher");
+#endif
   }
+
+  if (delegate_ != nullptr) {
+    if (interpreter->ModifyGraphWithDelegate (delegate_) != kTfLiteOk) {
+      ml_loge ("Failed to allocate tensors with NNAPI delegate\n");
+      return -2;
+    }
+  } else {
+    if (interpreter->AllocateTensors () != kTfLiteOk) {
+      ml_loge ("Failed to allocate tensors\n");
+      return -2;
+    }
+  }
+
 #if (DBG)
   gint64 stop_time = g_get_monotonic_time ();
   g_message ("Model is loaded: %" G_GINT64_FORMAT, (stop_time - start_time));

